### PR TITLE
Handle named tuples

### LINKED ISSUE #15: No completions on namedtuple
https://github.com/Microsoft/PTVS/issues/731
https://github.com/Microsoft/vscode-python/issues/2377

```python
from collections import namedtuple
RequestStartLine = namedtuple("RequestStartLine", "method path version".split())
start_line = RequestStartLine("GET", "/query", "HTTP/1.1")
print(start_line)
start_line.  # <- no suggestion
```
This needs new specializations

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Fixes #15

- Requires #304 so it is included here
- Adds handling of generic `Type[T]` from `typing` to `typeinfo` and the AST analysis
- Adds handling of `namedtuple` from `collections` via typeshed annotations
- Adds fetching of tuple members and handling assignment of tuple to sequence
- Adds specific 'class' flag to the Python type telling that function returns class rather than instance and hence calling it creates instance of the class rather than invoking on instance.
- Filters function overloads preferring annotated definitions to plain ones
- Relevant tests

Implement wrapper over tuple so we can expose declared members in completion - separate item in https://github.com/Microsoft/python-language-server/issues/343

### CODE REVIEW

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstAnalysisFunctionWalkerSet.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        public void ProcessFunction(FunctionDefinition fn) {
+            if (_functionWalkers.TryGetValue(fn, out var w)) {
+                ProcessWalker(w);
+            }
+        }
+
+        private void ProcessWalker(AstAnalysisFunctionWalker walker) {
+            // Remove walker before processing as to prevent reentrancy.
+            _functionWalkers.Remove(walker.Target);
+            var z = walker.Target.Name == "day";
```
Comment (MikhailArkhipov) [inline src/Analysis/Engine/Impl/Interpreter/Ast/AstAnalysisFunctionWalkerSet.cs]: Debug code, will remove

File: src/Analysis/Engine/Test/AstAnalysisTests.cs [hunk 2]
```diff
@@ -212,6 +191,50 @@ private static Task<Server> CreateServerAsync(InterpreterConfiguration configura
...
+                var analysis = await server.OpenDefaultDocumentAndGetAnalysisAsync(@"
+from ForwardRefFunc1 import *
+x = B().getA().methodA()
+");
+                analysis.Should().HaveVariable("x").OfTypes(BuiltinTypeId.Str);
+            }
+        }
+
+        [TestMethod, Priority(0)]
+        public async Task AstForwardRefFunction2() {
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Test/AstAnalysisTests.cs]: Can be a parameterized test.

File: src/Analysis/Engine/Impl/Values/BuiltinFunctionInfo.cs [hunk 3]
```diff
@@ -121,26 +122,10 @@ public override IAnalysisSet Call(Node node, AnalysisUnit unit, IAnalysisSet[] a
...
+        public override IEnumerable<OverloadResult> Overloads
+            => GetFunctionOverloads().Select(o => new BuiltinFunctionOverloadResult(ProjectState, Function.Name, o, 0, () => Description));
 
-        public override string Documentation {
-            get {
-                _doc = _doc ?? Utils.StripDocumentation(Function.Documentation);
-                return _doc;
-            }
-        }
+        public override string Documentation => _doc = _doc ?? (_doc = Utils.StripDocumentation(Function.Documentation));
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Values/BuiltinFunctionInfo.cs]: `public override string Documentation => _doc ?? (_doc = Utils.StripDocumentation(Function.Documentation));` should be enough

File: src/Analysis/Engine/Impl/Values/BuiltinFunctionInfo.cs [hunk 4]
```diff
@@ -25,27 +23,30 @@
...
         public override IPythonType PythonType => _type;
 
-        public override bool IsOfType(IAnalysisSet klass) {
-            return klass.Contains(ProjectState.ClassInfos[BuiltinTypeId.Function]) ||
-                klass.Contains(ProjectState.ClassInfos[BuiltinTypeId.BuiltinFunction]);
+        public override bool IsOfType(IAnalysisSet klass) 
+            => klass.Contains(ProjectState.ClassInfos[BuiltinTypeId.Function]) || klass.Contains(ProjectState.ClassInfos[BuiltinTypeId.BuiltinFunction]);
+
+        public override IAnalysisSet Call(Node node, AnalysisUnit unit, IAnalysisSet[] args, NameExpression[] keywordArgNames) {
+            var returnTypes = GetFunctionOverloads().Where(o => o.ReturnType != null).SelectMany(o => o.ReturnType);
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Values/BuiltinFunctionInfo.cs]: Do we need `.Distinct()` here?

File: src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs [hunk 5]
```diff
@@ -158,7 +159,22 @@ public BuiltinClassInfo(IPythonType classObj, PythonAnalyzer projectState)
         }
 
         public override IAnalysisSet GetIndex(Node node, AnalysisUnit unit, IAnalysisSet index) {
-            // TODO: Needs to actually do indexing on type
+            try {
+                var indices = index.Select(x => x.GetConstantValue()).ExcludeDefault().OfType<int>().ToArray();
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs]: `OfType<int>()` will remove defaults, so `ExcludeDefault()` isn't needed.

File: src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs [hunk 6]
```diff
@@ -158,7 +159,22 @@ public BuiltinClassInfo(IPythonType classObj, PythonAnalyzer projectState)
...
+                    var values= indices.Where(i => i >= 0 && i < a.Length).Select(i => unit.State.GetAnalysisValueFromObjects(a[i]));
+                    return AnalysisSet.UnionAll(values);
+                }
+                if (this is IBuiltinSequenceClassInfo bseqt) {
+                    var a = bseqt.IndexTypes.ToArray();
+                    var values = indices.Where(i => i >= 0 && i < a.Length).Select(i => a[i]);
+                    return AnalysisSet.UnionAll(values);
+                }
+            } catch {
+                return AnalysisSet.Empty;
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs]: What are we catching here?

File: src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs [hunk 7]
```diff
@@ -158,7 +159,22 @@ public BuiltinClassInfo(IPythonType classObj, PythonAnalyzer projectState)
         }
 
         public override IAnalysisSet GetIndex(Node node, AnalysisUnit unit, IAnalysisSet index) {
-            // TODO: Needs to actually do indexing on type
+            try {
+                var indices = index.Select(x => x.GetConstantValue()).ExcludeDefault().OfType<int>().ToArray();
+                if (this is IPythonSequenceType seqt) {
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs]: Is it possible? What type is derived from `BuiltinClassInfo` and implements `IPythonSequenceType`?

File: src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs [hunk 8]
```diff
@@ -158,7 +159,22 @@ public BuiltinClassInfo(IPythonType classObj, PythonAnalyzer projectState)
...
+                    var a = seqt.IndexTypes.ToArray();
+                    var values= indices.Where(i => i >= 0 && i < a.Length).Select(i => unit.State.GetAnalysisValueFromObjects(a[i]));
+                    return AnalysisSet.UnionAll(values);
+                }
+                if (this is IBuiltinSequenceClassInfo bseqt) {
+                    var a = bseqt.IndexTypes.ToArray();
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs]: Shouldn't we move this implementation into `GetIndex` override in `SequenceBuiltinClassInfo`?

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstTypeAnnotationConverter.cs [hunk 9]
```diff
@@ -196,6 +192,11 @@ class AstTypeAnnotationConverter : TypeAnnotationConverter<IPythonType> {
             return res;
         }
 
+        private IPythonType MakeGenericClassType(IPythonType typeArg)
+            => typeArg.IsBuiltin
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Interpreter/Ast/AstTypeAnnotationConverter.cs]: `UnionType` always return `IsBuiltin => false`. Is that ok?

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstTypeAnnotationConverter.cs [hunk 10]
```diff
@@ -86,21 +86,17 @@ class AstTypeAnnotationConverter : TypeAnnotationConverter<IPythonType> {
...
-        }
+        public override IPythonType MakeUnion(IReadOnlyList<IPythonType> types) => new UnionType(types);
 
-        public override IReadOnlyList<IPythonType> GetUnionTypes(IPythonType unionType) {
-            return (unionType as UnionType)?.Types ??
-                   (unionType as IPythonMultipleMembers)?.Members.OfType<IPythonType>().ToArray();
-        }
+        public override IReadOnlyList<IPythonType> GetUnionTypes(IPythonType unionType) =>
+            (unionType as UnionType)?.Types ??
+            (unionType as IPythonMultipleMembers)?.Members.OfType<IPythonType>().ToArray();
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Interpreter/Ast/AstTypeAnnotationConverter.cs]: A bit hard to read. What would you say about changing this to:
```cs
type is UnionType unionType
    ? unionType.Types:
    : type is IPythonMultipleMembers multipleMembers
        ? multipleMembers.Members.OfType<IPythonType>().ToArray()
        : null;
```

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonType.cs [hunk 11]
```diff
@@ -232,8 +236,8 @@ ILocationInfo loc
 
         public IEnumerable<string> GetMemberNames(IModuleContext moduleContext) {
             var names = new HashSet<string>();
-            lock (_members) {
-                names.UnionWith(_members.Keys);
+            lock (Members) {
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonType.cs]: I really don't like an idea of taking locks on mutable, non-private field. `AstPythonBuiltinType` needs it only inside `IsHidden`, which can be fixed by adding
```cs
protected bool ContainsMember(string name) {
    lock (_members) {
        return Members.ContainsKey(name);
    }
}
```

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonProperty.cs [hunk 12]
```diff
@@ -5,34 +5,31 @@
...
 
 namespace Microsoft.PythonTools.Interpreter.Ast {
-    class AstPythonProperty : IBuiltinProperty, ILocatedMember {
+    class AstPythonProperty : IBuiltinProperty2, ILocatedMember {
         private IPythonFunctionOverload _getter;
 
         public AstPythonProperty(
             PythonAst ast,
-            FunctionDefinition getter,
+            FunctionDefinition definition,
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonProperty.cs]: 👍

File: src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs (hunk 7 above)
Comment (MikhailArkhipov) [reply src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs]: Looks like no one

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstTypeAnnotationConverter.cs (hunk 9 above)
Comment (MikhailArkhipov) [reply src/Analysis/Engine/Impl/Interpreter/Ast/AstTypeAnnotationConverter.cs]: I changed implementation, will update soon

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonType.cs (hunk 11 above)
Comment (MikhailArkhipov) [reply src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonType.cs]: OK

File: src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs (hunk 8 above)
Comment (MikhailArkhipov) [reply src/Analysis/Engine/Impl/Values/BuiltinClassInfo.cs]: Yes

File: src/Analysis/Engine/Impl/AnalysisValueSetExtensions.cs [hunk 13]
```diff
@@ -140,8 +135,7 @@ public static class AnalysisValueSetExtensions {
         /// Performs a delete index operation propagating the index types into
         /// the provided object.
         /// </summary>
-        public static void DeleteIndex(this IAnalysisSet self, Node node, AnalysisUnit unit, IAnalysisSet index) {
-        }
+        public static void DeleteIndex(this IAnalysisSet self, Node node, AnalysisUnit unit, IAnalysisSet index) { }
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/AnalysisValueSetExtensions.cs]: Empty static method?

Comment (MikhailArkhipov) [reply src/Analysis/Engine/Impl/AnalysisValueSetExtensions.cs]: Apparently it is called from DDG... Let's leave this alone until we decide fate of the DDG.

**Pull request merged**